Kernel/Devices/VMWareSVGADevice.cpp

```diff
@@ -0,0 +1,271 @@
+    m_fifo_addr = m_fifo_mapping->vaddr();
+    // Clear framebuffer memory
+    OwnPtr<Region> clear_map = MM.allocate_kernel_region(m_framebuffer_address_physical, page_round_up(PCI::get_BAR_space_size(address, 1)), "VMWareSVGA FB Test", Region::Access::Write, Region::Cacheable::No);
+    memset(clear_map->vaddr().as_ptr(), 0x00, PCI::get_BAR_space_size(address, 1));
+    clear_map->unmap();
+    VERIFY(m_io_base);
+    VERIFY(m_framebuffer_address_physical.as_ptr());
+    VERIFY(m_fifo_addr.as_ptr());
+
+    // Determine SVGA Version
+    // The way this is done is by writing the highest version we support (2) into a device support register,
+    // and checking if the value has been changed.
+    // If it hasn't, then we found the highest version that both we and the "card" support.
+    // If it has, decrement by one and repeat.
+    m_device_version = make_version_id(2);
+    set_register(IORegister::ID, m_device_version);
+    while (get_register(IORegister::ID) != m_device_version && m_device_version >= make_version_id(0)) {
+        m_device_version--;
+        set_register(IORegister::ID, m_device_version);
+    }
+
+    VERIFY(m_device_version >= make_version_id(0));
+
+    // We want to read the size of the Fifo memory
+    m_fifo_size = get_register(IORegister::MemSize);
+
+    // The reference driver checks some sizes here, but Qemu actually makes its buffers for
+    // the fifo smaller then VMWare does, so they fail. The fifo still works, however, and the
+    // numbers are accurate.
+
+    // Check if the card is new enough and get the capabilities register
+    if (m_device_version >= make_version_id(1)) {
+        m_device_capabilities = get_register(IORegister::Capabilities);
+    }
+
+    // Check if interrupt masks are supported
+    if (get_capability(Capabilities::IRQMask)) {
+        // Get the card interrupt line
+        m_interrupt_line = PCI::get_interrupt_line(address);
+        change_irq_number(m_interrupt_line);
+        // Mask all interrupts
+        set_register(IORegister::IRQMask, 0xFF);
+        // Clear all interrupts on the card
+        IO::out32(m_io_base + static_cast<u32>(Ports::IRQStatus), 0xFF);
+        // Enable the interrupt handler
+        enable_irq();
+    }
+
+    dbgln("VMWareSVGA: handshake complete, version {}, max res: {}x{}, irq: {}", m_device_version & 0xFF, get_register(IORegister::MaxWidth), get_register(IORegister::MaxHeight), m_interrupt_line);
+
+    // FIXME: Remove this when we get signals for updates from WindowServer
+    RefPtr<Thread> update_thread;
+    Process::create_kernel_process(update_thread, "VMWareSVGAThread", [&] {
+        dbgln("VMWareSVGA: Thread is running");
+        for (;;) {
+            force_vblank();
+            (void)Thread::current()->sleep(Time::from_milliseconds(10));
```

Comment:
This suggests 100 updates per second. Might be better to refresh ~60 times a second instead which avoids some extra work.
```suggestion
            (void)Thread::current()->sleep(Time::from_milliseconds(16));
```
(This isn't as important since this is temporary code.)

Reply:
Since screen updates are, as far as I know, atomic, this doesn't really matter as long as it's below the refresh rate. One might even want to make it slightly less than 16ms, as it likely won't get perfectly synced with the actual refresh rate.